File: libs/core/uart.cpp
#include "pxt.h"

// Native UART for the `uart` shims (uart.ts). This is a SECOND serial bus, separate
// from console.log (which uses USB + the stdio UART). The RP2040 UART peripheral
// (uart0/uart1) is chosen from the TX pin. Host VM build logs via DMESG.

#ifdef PXT_RP2040
#include "hardware/uart.h"
#include "hardware/gpio.h"

// RP2040 pinmux for UART (by 4-pin group): GP0-3,12-19,28 -> uart0 ; GP4-11,20-27 -> uart1.
// group = gpio/4 ; instance = ((group+1)/2) & 1.
static uart_inst_t *uart_for(int gpio) {
    int group = gpio >> 2;
    return (((group + 1) >> 1) & 1) ? uart1 : uart0;
}

static uart_inst_t *uartDev;
static int uartBaud = 9600;
#endif

namespace uart {

//%
void init(int tx, int rx, int baud) {
#ifdef PXT_RP2040
    uartBaud = baud;
    uartDev = uart_for(tx);
    uart_init(uartDev, baud); // 8N1 by default
    gpio_set_function(tx, GPIO_FUNC_UART);
    gpio_set_function(rx, GPIO_FUNC_UART);
#else
    DMESG("uart.init tx=%d rx=%d baud=%d", tx, rx, baud);
#endif
}

//%
void setBaudRate(int baud) {
#ifdef PXT_RP2040
    uartBaud = baud;
    if (uartDev)
        uart_set_baudrate(uartDev, baud);
#else
    DMESG("uart.setBaudRate %d", baud);
#endif
}

//%
void writeBuffer(Buffer buf) {
#ifdef PXT_RP2040
    if (uartDev && buf && buf->length > 0)
        uart_write_blocking(uartDev, buf->data, buf->length);
#else
    DMESG("uart.writeBuffer len=%d", buf ? buf->length : 0);
#endif
}

//%
Buffer readBuffer(int size) {
    if (size < 0)
        size = 0;
#ifdef PXT_RP2040
    Buffer buf = mkBuffer(NULL, size);
    if (uartDev && size > 0)
        uart_read_blocking(uartDev, buf->data, size);
    return buf;
#else
    DMESG("uart.readBuffer size=%d", size);
    return mkBuffer(NULL, size);
#endif
}

//%
int available() {
#ifdef PXT_RP2040
    if (!uartDev)
        return 0;
    return uart_is_readable(uartDev) ? 1 : 0;
#else
    return 0;
#endif
}

} // namespace uart

File: libs/core/uart.ts
/**
 * UART / serial communication on a second hardware UART (separate from console.log,
 * which uses USB + the stdio UART).
 */
//% color="#B22222" weight=86 icon="" block="UART"
namespace uart {
    let _inited = false;

    function ensure(): void {
        if (!_inited) {
            // default pins: TX=GP8, RX=GP9 (UART1) so it doesn't clash with the
            // console's UART0 on GP0/1
            init(DigitalPin.P8, DigitalPin.P9, 9600);
        }
    }

    /**
     * Initialize the UART on the given pins and baud rate.
     * The hardware UART peripheral is chosen automatically from the TX pin.
     * @param tx the transmit pin
     * @param rx the receive pin
     * @param baud baud rate, eg: 9600
     */
    //% blockId=uart_init block="UART init tx %tx|rx %rx|baud %baud" weight=100
    export function init(tx: DigitalPin, rx: DigitalPin, baud: number = 9600): void {
        __init(tx, rx, baud);
        _inited = true;
    }

    /**
     * Set the baud rate.
     * @param baud baud rate, eg: 9600
     */
    //% blockId=uart_baud block="UART set baud rate %baud" weight=90
    export function setBaudRate(baud: number): void {
        ensure();
        __setBaudRate(baud);
    }

    /**
     * Write a string to the UART.
     */
    //% blockId=uart_write_string block="UART write string %text" weight=85
    export function writeString(text: string): void {
        ensure();
        __writeBuffer(control.createBufferFromUTF8(text));
    }

    /**
     * Write a buffer to the UART.
     */
    //% blockId=uart_write_buffer block="UART write buffer %buf" weight=84
    export function writeBuffer(buf: Buffer): void {
        ensure();
        __writeBuffer(buf);
    }

    /**
     * Read exactly `size` bytes from the UART (blocking).
     * @param size number of bytes to read
     */
    //% blockId=uart_read_buffer block="UART read %size|bytes" weight=80
    export function readBuffer(size: number): Buffer {
        ensure();
        return __readBuffer(size);
    }

    /**
     * Number of bytes available to read without blocking (0 or more).
     */
    //% blockId=uart_available block="UART available" weight=75
    export function available(): number {
        ensure();
        return __available();
    }
}

declare namespace uart {
    //% shim=uart::init
    function __init(tx: DigitalPin, rx: DigitalPin, baud: int32): void;
    //% shim=uart::setBaudRate
    function __setBaudRate(baud: int32): void;
    //% shim=uart::writeBuffer
    function __writeBuffer(buf: Buffer): void;
    //% shim=uart::readBuffer
    function __readBuffer(size: int32): Buffer;
    //% shim=uart::available
    function __available(): int32;
}
